pkg/cmd/admin/migrate/images/images.go

```diff
@@ -0,0 +1,488 @@
+			kcmdutil.CheckErr(options.Validate())
+			if err := options.Run(); err != nil {
+				// TODO: move met to kcmdutil
+				if err == cmdutil.ErrExit {
+					os.Exit(1)
+				}
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+	options.ResourceOptions.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateImageReferenceOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	var remainingArgs []string
+	for _, s := range args {
+		if !strings.Contains(s, "=") {
+			remainingArgs = append(remainingArgs, s)
+			continue
+		}
+		mapping, err := ParseMapping(s)
+		if err != nil {
+			return err
+		}
+		o.Mappings = append(o.Mappings, mapping)
+	}
+
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c, remainingArgs); err != nil {
+		return err
+	}
+
+	osclient, _, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Client = osclient
+
+	return nil
+}
+
+func (o *MigrateImageReferenceOptions) Validate() error {
+	if len(o.Mappings) == 0 {
+		return fmt.Errorf("at least one mapping argument must be specified: REGISTRY/NAME=REGISTRY/NAME")
+	}
+	return o.ResourceOptions.Validate()
+}
+
+func (o *MigrateImageReferenceOptions) Run() error {
+	return o.ResourceOptions.Visit(func(info *resource.Info) (migrate.Reporter, error) {
+		return o.transform(info.Object)
+	})
+}
+
+// save invokes the API to alter an object. The reporter passed to this method is the same returned by
+// the migration visitor method (for this type, transformImageReferences). It should return an error
+// if the input type cannot be saved. It returns migrate.ErrRecalculate if migration should be re-run
+// on the provided object.
+func (o *MigrateImageReferenceOptions) save(info *resource.Info, reporter migrate.Reporter) error {
+	switch t := info.Object.(type) {
+	case *imageapi.ImageStream:
+		if reporter.(imageChangeInfo).status {
+			updated, err := o.Client.ImageStreams(t.Namespace).UpdateStatus(t)
+			if err != nil {
+				return migrate.DefaultRetriable(info, err)
+			}
+			info.Refresh(updated, true)
+			return migrate.ErrRecalculate
+		}
+		if reporter.(imageChangeInfo).spec {
+			updated, err := o.Client.ImageStreams(t.Namespace).Update(t)
+			if err != nil {
+				return migrate.DefaultRetriable(info, err)
+			}
+			info.Refresh(updated, true)
+		}
+		return nil
+	default:
+		if _, err := resource.NewHelper(info.Client, info.Mapping).Replace(info.Namespace, info.Name, false, info.Object); err != nil {
+			return migrate.DefaultRetriable(info, err)
+		}
+		//return fmt.Errorf("resource %q does not have a save method implemented (%T)", info.Mapping.Resource, t)
+	}
+	return nil
+}
+
+// transform checks image references on the provided object and returns either a reporter (indicating
+// that the object was recognized and whether it was updated) or an error.
+func (o *MigrateImageReferenceOptions) transform(obj runtime.Object) (migrate.Reporter, error) {
+	fn := o.Mappings.MapReference
+	switch t := obj.(type) {
```

Comment:
I think this structure needs to allow for inspection of which types are handled, similar to how we handle the node selector admission plugin so we can be sure that we're inspecting at least those types that we know need updating.  In fact, I think you two are pulling the same field out.  It'd be nice to fix up that TODO at the same time: https://github.com/openshift/origin/blob/master/pkg/scheduler/admission/podnodeconstraints/admission.go#L65-L67.  Something like `map[reflect.Type]func(runtime.Object) *PodSpec` ought to do it, right?